### DISCUSSION TITLE
run `bin` shim while using `shortcut` for jar files

### DISCUSSION THREAD

**Discussion by goyalyashpal** [ScoopInstaller/Scoop#5873] [Help and Questions] [answered]
titles:
* run `bin` shim while using `shortcut` for jar files
* access external programs in bin/shortcuts for running scripts

hi!

* there are lots of script like programs which run of some interpreter or some external program
* example: `.jar` files run off of `java.exe -jar`, `.ahk` files run via ahk, `.py` projects often run off of `pip` or `pdm` etc.

i tried specifying like so in the manifest file, and on trying to install, i get shown an error like so:

```json
    "bin": [ [ "java.exe", "program", "-jar program.jar" ] ], 
    "shortcuts": [ [ "java.exe", "Program", "-jar program.jar" ] ],
```

```
Creating shim for 'program'.
Creating shortcut for Program (java.exe) failed: Couldn't find ...\scoop\apps\program\current\java.exe
```

i can understand why this behaviour is currently implemented like so. but is there any way to accomplish for this use case?

earlier, i had manually gathered those source file to local storage, and had created a manual `scoop shim` like so:

```console
$ history | grep "scoop shim add"
scoop shim add program 'java' -- -jar "...\program-1.2.3.jar" --console
scoop shim add program 'java' -- -jar "...\program-1.2.3.jar"

$ scoop shim info program
Name     : program
Path     : ...\scoop\shims\program.exe
Source   : temurin-jdk
Type     : Application
IsGlobal : False
IsHidden : False

$ bat -p $(!! | yq .Path | sd "exe$" "shim")
path = "...\scoop\apps\temurin-jdk\current\bin\java.exe"
args = -jar program.jar
```

### versions and os:

```console
$ scoop --version | head -n2
Current Scoop version:
v0.3.1 - Released at 2022-11-15

$ scoop config
root_path    : ...\scoop
global_path  : ...\scoop_global
last_update  : 2024-04-11T07:19:06.4988281+00:00
alias        : @{cat-dum=scoop-cat-dum}
debug        : True
scoop_branch : master
scoop_repo   : https://github.com/ScoopInstaller/Scoop

$ systeminfo | grep -iA1 "os name"
OS Name:                   Microsoft Windows 10 Home Single Language
OS Version:                10.0.19045 N/A Build 19045

```

**Comment by niheaven** [comment, marked as answer]
Just use `xx.jar` or `xx.py` as bin, since scoop will automatically add `java` or `python` when creating shims. See manifests in https://github.com/search?q=repo%3AScoopInstaller%2FMain%20.jar&type=code as reference.

**Reply by goyalyashpal** [reply]
holllly molly. that is also a thing in scoop. that's suppper neat and thoughtful.

it `"bin": "program.jar"` works fine as `bin` 😃😇 
is there any trick to do so for `shortcuts` i.e. GUI like thingies too 😄 ?

explanation: `program.jar` works as GUI application, `program.jar --console` or `program.jar --cli` etc works on cli

**Reply by niheaven** [reply]
Hmm, it seems that you should create a bat yourself, there's some examples in extras bucket, ref https://github.com/search?q=repo%3AScoopInstaller%2FExtras%20.jar&type=code

**Reply by goyalyashpal** [reply]
well, that might work. but in this case, all that the shortcut would need is to be able to directly run the bin-shim as it is... nothing else.

why do even that and not use the bin-shim itself u may ask. 'cz the bin-shim requires entering full name of command when starting from start menu. whereas a shortcut would appear in autocomplete suggestions too
